### PR TITLE
Migrate nodeprocessor and its minor dependencies to typescript

### PR DESCRIPTION
**Pull request by ang-zeyu**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Tackles part of #1913 

**Overview of changes:**
- Migrate the entrypoint to the `src/html` folder, `NodeProcessor`
- Some of its minor dependencies referenced (e.g. `Context` class)

This should enable migration of the rest of the `src/html` folder more smoothly/independently

**Anything you'd like to highlight/discuss:**
See commit message

**Testing instructions:**
na

**Proposed commit message: (wrap lines at 72 characters)**

NA (in the commits themselves), let's do a rebase here to preserve the file rename

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:119]: I feel like there should be better ways to configure the typing here, let me know if any ideas 🙂 

Or we can defer to another PR to investigate, since functionally it is the same.

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:140]: had to change this part functionally since `this[property]` is not going to pass tsc.

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:147]: only change here is this. This lets tsc know `node.attribs` (which is defined as `?: {}`) always exists

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:167]: same hint to tsc.

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:342]: a more horrendous typing issue with cheerio/htmlparser2. Functionally the same but dosen't look pretty.

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:363]: same sort of hint to tsc to let it know `node.name` always exists.

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:370]: Added `?.` here (and line 365)

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [inline packages/core/src/html/NodeProcessor.ts:408]: Another functional change. But `context` is unused inside `processNode` here.

For standardising the typing of `processNode` here (2nd parameter is nullable). The other option is to make the 2nd parameter of `processNode(..., context?: Context)`, but that requires some hinting to tsc about whether `context` exists. This seemed like the cleaner option.

File: packages/core/src/html/NodeProcessor.ts
Comment (jonahtanjz) [inline packages/core/src/html/NodeProcessor.ts:406]: Looks like there may be some typing issue here? `mainHtmlNodes` here has the type `(DomElement | DomElement[])[]` which makes `d` have the type `DomElement | DomElement[]`. `_trimNodes` however only accepts a `DomElement` as param.

Also seems like `_trimNodes` method will have some issues if an array of `DomElement`s is passed to it.

File: packages/core/src/html/NodeProcessor.ts
Comment (jonahtanjz) [inline packages/core/src/html/NodeProcessor.ts:398]: `this.traverse` can return a `DomElement | DomElement[]`

File: packages/core/src/html/NodeProcessor.ts
Comment (jonahtanjz) [inline packages/core/src/html/NodeProcessor.ts:402]: I think you can update this to `import { createErrorNode } from './elements'` since `element` has been converted to typescript. This should also let typescript check the typing of `mainHtmlNodes` which will cause the error.

File: packages/core/src/html/NodeProcessor.ts
Comment (jonahtanjz) [inline packages/core/src/html/NodeProcessor.ts:309]: From what I can see, it seems like this check is not necessary since `node` will not be an array? If that is the case, we can remove this block and `traverse` should just return a `DomElement`.

**Review by jonahtanjz — COMMENTED** [review]
Thanks @ang-zeyu 👍 

Just one concern, other than that the rest looks good :)

File: packages/core/src/html/NodeProcessor.ts
Comment (ang-zeyu) [reply packages/core/src/html/NodeProcessor.ts:389]: Nice catch, changed to just `DomElement`. Also resolves the other issue you pointed out with `_trimNodes`

**Comment by ang-zeyu** [comment]
Changes are in the two fixups, have rebased on the latest master branch.

I also found a solution for the horrible typing issue [here](https://github.com/MarkBind/markbind/pull/2029#discussion_r962142464).
 Just left with this one https://github.com/MarkBind/markbind/pull/2029#discussion_r962142184 if you have any ideas. =C

File: packages/core/src/html/NodeProcessor.ts
Comment (jonahtanjz) [reply packages/core/src/html/NodeProcessor.ts:120]: I guess we can defer this to another PR to investigate. Can't think of a good idea right now...

**Review by jonahtanjz — APPROVED** [review]
LGTM 👍

**Comment by ang-zeyu** [comment]
Hi @jonahtanjz! Didn't expect you to merge in so quickly. It's been very long since we had a pr requiring a[ fixup workflow](https://dev.to/koffeinfrei/the-git-fixup-workflow-386d), sorry for not reminding you of the intended process earlier.

It usually goes like this (the fixups should be squashed into the main commits once reviewed; only the main commits should be merged).
This PR isn't very big, but we're using a similar workflow for preserving renames.

```
1. Contributor sends up a rather big PR that should have a multi commit structure
  - If it is already structured, skip to step 4
2. Maintainer informs contributor its too big for one commit 
3. Contributor soft resets git branch, recommits and pushes with structure
4. The PR is reviewed
5. Contributor acts on the review by delivering "fix ups" (**incremental** commits that are reviewable, each one tied to a separate commit in the structure earlier) (https://blog.sebastian-daschner.com/entries/git-commit-fixup-autosquash)
6. Reviewers review these fixups
  - All good -> step 9, otherwise step 7
7. Contributor does an interactive rebase / uses the git fixup feature to squash each fixup commit earlier with its respective commit, and adds **new** fixups to tackle step 6's issues.
8. Go back to step 6
9. Maintainer merges using a **merge commit**
```

The intention is to keep a bunch of well-organized, semantic commits throughout the review, and after merge, but also make incremental changes easy to review (via fixups).

**Comment by ang-zeyu** [comment]
My opinion is that its fine to leave some derps in the commit history; But if you wish, you can also overwrite the relevant parts with an interactive rebase, since its fairly recent, and the changes once squashed are identical. (shouldn't be any issues with merging other prs on top)

**Pull request merged**
